zkvm/src/signature/signature.md

```diff
@@ -0,0 +1,253 @@
+- aggregated_key: `VerificationKey`
+
+Functions: 
+- `Multikey::new(...)`: detailed more in [key aggregation](#key-aggregation) section. 
+- `Multikey::factor_for_key(self, verification_key)`: computes the `a_i` factor, where `a_i = H(<L>, X_i)`. `<L>`, or the list of pukeys that go into the aggregated pubkey, has already been committed into `self.transcript`. Therefore, this function simply clones `self.transcript`, commits the verification key (`X_i`) into the transcript with label "X_i", and then squeezes the challenge scalar `a_i` from the transcript with label "a_i".
+- `Multikey::aggregated_key(self)`: returns the aggregated key stored in the multikey, `self.aggregated_key`.  
+
+### Signature
+
+A signature is comprised of a scalar `s`, and a RistrettoPoint `R`. In the simple Schnorr signature case, `s` represents the Schnorr signature scalar and `R` represents the nonce commitment. In the MuSig signature case, `s` represents the sum of the Schnorr signature scalars of each party, or `s = sum_i (s_i)`. `R` represents the sum of the nonce commitments of each party, or `R = sum_i (R_i)`. 
+
+Functions:
+- `Signature::verify(...)`: In both the simple Schnorr signature and the MuSig signature cases, the signature verification is the same. For more detail, see the [verification](#verifying) section.
+
+## Operations
+
+### Key Aggregation
+
+Key aggregation happens in the `Multikey::new(...)` function. 
+
+Input:
+- pubkeys: `Vec<VerificationKey>`. This is a list of compressed public keys that will be aggregated, as long as they can be decompressed successfully.
+
+Operation:
+- Create a new transcript using the tag "ZkVM.aggregated-key". (TODO: remove the "ZkVM." if we make the signature crate separate from the ZkVM crate.)
+- Commit all the pubkeys to the transcript. This reflects the creation of the `<L>` encoding of the list of pubkeys detailed in the MuSig paper: `<L> = H(X_1 || X_2 || ... || X_n)`. The difference is that we do not need to squeeze `<L>` out of the transcript, since we can use the prepared transcript directly for future randomness.
+  (TODO: this sounds awkward, explain better?)
+- Create `aggregated_key = sum_i ( a_i * X_i )`. Iterate over the pubkeys, compute the factor `a_i = H(<L>, X_i)`, and add `a_i * X_i` to the aggregated key.
+
+Output:
+- a new multikey, with the transcript and aggregated key detailed above.
+
+### Signing
+
+There are several paths to signing:
+1. Make a Schnorr signature with one public key (derived from one private key).
+    Function: `Signature::sign_single(...)`
+
+    Input: 
+    - transcript: `&Transcript` - the message to be signed should have been committed to the transcript beforehand. (TODO: pass in a mutable borrow of a transcript instead of a borrow.)
+    - privkey: `Scalar`
+
+    Operation:
+    - Use the transcript to generate a random factor (the nonce, `r`), by committing to the privkey and passing in a `thread_rng`
```

Comment:
we need to clone transcript before mixing it with privkey and RNG.

```suggestion
    - Clone the transcript state, mix it with the privkey and system-provided RNG to generate the nonce `r`. This makes the nonce uniquely bound to a message and private key, and also makes it non-deterministic to prevent "rowhammer" attacks.
```